game/menu/gamemanager.gd

```diff
@@ -13,22 +20,165 @@ func _ready():
+
+
+## Load data from a file or GAME_DATA_CACHE and return the data for the game.
+## If something doesn't exist an empty Dictionary is returned, which is put in
+## the correct location in the GAME_DATA_CACHE Dictionary.
+## This is a private function, don't use this in a game.
+func _load_data(file_name: String, game: String) -> Dictionary:
+	var cache_key = get_current_player() + "/" + game + "/" + file_name
+	var directory = "user://" + get_current_player() + "/" + game + "/"
+	Directory.new().make_dir_recursive(directory)
+	if cache_key in GAME_DATA_CACHE:
+		return GAME_DATA_CACHE[cache_key]
+
+	GAME_DATA_CACHE[cache_key] = {}  # populate with default
+	# read saved data from a file
+	var file = File.new()
+	file.open(directory + file_name + ".json", File.READ)
+
+	if not file.is_open():
+		return GAME_DATA_CACHE[cache_key]
+
+	var content = file.get_as_text()
+	file.close()
+
+	var parse_result = JSON.parse(content)
+	if not parse_result.error:
+		# put contents from file in the cache
+		GAME_DATA_CACHE[cache_key] = parse_result.result
+
+	return GAME_DATA_CACHE[cache_key]
+
+
+## Save the changes to the dict returned by get_game_data()
+## This method is automatically called when a game ends.
+func save_game_data():
+	assert(_last_loaded_game != "")  # this should only be called if _last_loaded_game is set
+	_save_data("game_data", _load_data("game_data", _last_loaded_game), _last_loaded_game)
+
+
+## Get the game data for the current player and the current game.
+## To save data. Just modify the returned Dictionary and call save_game_data().
+func get_game_data() -> Dictionary:
+	assert(_last_loaded_game != "")
+	return _load_data("game_data", _last_loaded_game)
+
+
+func get_current_player() -> String:
+	return "p"  # in future add here more logic
+
+
+## Get the time the current player played the currently running game last.
+func get_last_played(game_id = null):
+	game_id = _last_loaded_game if game_id == null else game_id
+	assert(game_id != "")
+	var data = _load_data("game_meta_data", game_id)
+	if not "last_played" in data:
+		return null
+	var dt = OS.get_datetime_from_unix_time(data["last_played"])
+	return (
+		"%04d-%02d-%02d %02d:%02d:%02d UTC"
+		% [dt["year"], dt["month"], dt["day"], dt["hour"], dt["minute"], dt["second"]]
+	)
+
+
+## Get time the current player played the currently running game.
+func get_played_time(game_id = null) -> float:
+	game_id = _last_loaded_game if game_id == null else game_id
+	assert(game_id != "")
+	var data = _load_data("game_meta_data", game_id)
+	if not "played_time" in data:
+		return 0.0
+	return data["played_time"] / 1000.0
+
+
+## Get the high_score of the current player for the currently running game.
+func get_high_score(game_id = null):
+	game_id = _last_loaded_game if game_id == null else game_id
+	if not game_id:  # game_id should be the folder_name, not null or ""
+		return null
+	var data = _load_data("game_meta_data", game_id)
+
+	if not "scores" in data:
+		return null
```

Comment:
Shouldn't a non-existing high-score automatically be `0`?

Reply:
I don't think so, what if games allow for a negative score? I think there should be a difference

Reply:
Does it even have to be a number? Maybe games could define a custom score class?

Reply:
Currently I think it has to be a number. I don't see how a custom score class would be any benefit.

Reply:
I think one way or another one needs to define a way to compare a score to others. Commonly this can be achieved with conversion to a number. But what about a game like a way out, where the level / difficulty might influence your score as well?

A Class might allow easier implementation of that. However for classic mini games a high score is more common in arcade games, right?

Also prefer null instead of 0 for the initial score.